main/src/library/Graph.flix

```diff
@@ -108,6 +139,49 @@ namespace Graph {
     /// Returns the shortest distance between all pairs of vertices in the
     /// weighted directed graph `g`.
     ///
+    pub def stableDistances(g: m[(t, Int32, t)]): Option[Map[(t, t), Int32]] with Foldable[m], Boxable[t] = {
+        // based on floyd-warshall
+        def rank(v, s) = s |> Set.partition(v0 -> v0 < v) |> fst |> Set.size;
+        let edges = inject g into Edge;
+        let negCycle = #{
+            Node(x) :- Edge(x, _, _).
+            Node(x) :- Edge(_, _, x).
+            // Collect the nodes to find the rank of each node (the number of
+            // nodes less than)
+            Nodes(;Set#{x}) :- Node(x).
+            NodeCount(Set.size(ns)) :- fix Nodes(;ns).
+            NodeRank(x, rank(x, ns)) :- Node(x), fix Nodes(;ns).
+            Dist(x, x, 0; Down(0)) :- Node(x).
+            Dist(x, y, 0; Down(w)) :- Edge(x, w, y).
+            Dist(x, y, k+1; w) :- Dist(x, y, k; w), NodeCount(c), if k < c.
+            Dist(x, y, k+1; w1 + w2) :- Dist(x, kth, k; w1), Dist(kth, y, k; w2),
+                NodeRank(kth, k), NodeCount(c), if k < c.
+            NegCycle() :- fix Dist(x, x, _; w), if destructDown(w) < 0.
```

Comment:
I worry about correctness. How small can `w` get? Can it underflow?

Reply:
It is bounded by `Dist(_, _, k; _)`. You can at most take n steps. It overflows if there is any shortest path for which the distance overflows. This is the same as distancesFrom and those related ones